src/libcore/ffi.rs

```diff
@@ -369,5 +420,96 @@ extern "rust-intrinsic" {
     /// Loads an argument of type `T` from the `va_list` `ap` and increment the
     /// argument `ap` points to.
     #[cfg(not(bootstrap))]
+    #[cfg(not(all(target_arch = "aarch64", not(target_os = "ios"), not(windows))))]
     fn va_arg<T: sealed_trait::VaArgSafe>(ap: &mut VaListImpl<'_>) -> T;
 }
+
+#[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+use crate::{mem, ptr};
+
+// Implemenation for AArch64 linux (and others that follow the standard PCS)
+// based on https://developer.arm.com/docs/ihi0055/d/procedure-call-standard-for-the-arm-64-bit-architecture
+// APPENDIX Variable argument Lists -> The va_start() macro
+#[cfg(all(target_arch = "aarch64", not(target_os = "ios"), not(windows)))]
+impl<'f> VaListImpl<'f> {
+    unsafe fn va_arg_on_stack<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        let mut arg = self.stack as usize;
+        if mem::align_of::<T>() > 8 {
+            arg = (arg + 15) & !15;
+        }
+        self.stack = ((arg + mem::size_of::<T>() + 7) & !7) as *mut c_void;
+        #[cfg(target_endian = "big")]
+        {
+            if
+            /* classof(type) != "aggregate" && */
+            mem::size_of::<T>() < 8 {
+                arg += 8 - mem::size_of::<T>();
+            }
+        }
+        ptr::read(arg as *const T)
+    }
+
+    unsafe fn va_arg_gr<T: sealed_trait::VaArgSafe>(&mut self) -> T {
+        let mut offs = self.gr_offs;
+        if offs >= 0 {
+            return self.va_arg_on_stack(); // reg save area empty
+        }
+        if mem::align_of::<T>() > 8 {
+            offs = (offs + 15) & !15; // round up
+        }
+        let nreg = (mem::size_of::<T>() as i32 + 7) / 8;
+        self.gr_offs = offs + (nreg * 8);
+        if self.gr_offs > 0 {
+            return self.va_arg_on_stack(); // overflowed reg save area
+        }
+        #[cfg(target_endian = "big")]
+        {
+            if
+            /* classof(type) != "aggregate" && */
+            mem::size_of::<T>() < 8 {
+                offs += 8 - mem::size_of::<T>() as i32;
+            }
+        }
+        ptr::read((self.gr_top as usize).wrapping_add(offs as usize) as *const T)
+    }
+
+    // Rust doesn't support these yet
```

Comment:
Would moving this implementation into the compiler (inside [src/librustc_codegen_llvm/va_arg.rs](https://github.com/rust-lang/rust/blob/master/src/librustc_codegen_llvm/va_arg.rs) let you implement this?

Reply:
Actually, are these SIMD types? There are some of those in nightly, even for AArch64.

Reply:
Yeah it would and allow supporting aggregates in general. What I meant by this comment was rust doesn't support aggregates in variadic functions (yet) so it doesn't need to be implemented.